Add content only descriptions in dropdown menus for patterns and templates

kevin940726 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Based on #61012. Implement the design in https://github.com/WordPress/gutenberg/issues/60834#issuecomment-2063662703.

Add descriptions for content-only blocks inside patterns or templates in the dropdown menu.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

See #60834.

This is a UI enhancement for content-only blocks. Most of the actions in the original dropdown aren't needed for content-only blocks. We remove them and provide descriptions and links to edit the entities instead.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Hide original actions and inject descriptions for content-only blocks. The wording and the design are still TBD and can be tweaked.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
**For patterns**:
- Create a synced pattern with overridable blocks.
- Insert the pattern into a post.
- Use list view to select the overridable blocks and open the options dropdown.
- Expect to see the description and the "Edit pattern" menu item.
- Clicking on "Edit pattern" should navigate to the pattern editor.
- The same flow should also work in the site editor.

**For templates**:
- Enable twentytwentyfour theme.
- Go to Site editor -> Pages -> Privacy Policy.
- Open list view and select the "Featured Image" block.
- Open the Options dropdown and expect to see the description and the "Edit template" menu item.
- Clicking on it should navigate to the template editor.

## Screenshots or screencast <!-- if applicable -->

**Patterns**:
![pattern](https://github.com/WordPress/gutenberg/assets/7753001/ccf02911-f4ef-4899-b1cc-837a22d0b961)

**Templates**:
![template](https://github.com/WordPress/gutenberg/assets/7753001/b58cf628-5cf0-41f8-a675-d2b720a2975d)

kevin940726 committed:
Remove the lock icon block content only blocks

kevin940726 committed:
Add content only description for patterns and templates

kevin940726 commented:
> ![template](https://github.com/WordPress/gutenberg/assets/7753001/3661d973-1b7c-4395-bcea-62b065382d0e)

Do we still want the "Create template part" action in the above screenshot?

Also, I noticed that the "Content" block shows the "Move to" action and can be grabbed but can't actually be moved anywhere. Not sure if this is a bug or not.

![image](https://github.com/WordPress/gutenberg/assets/7753001/05c670e3-9eb8-474e-ab63-1a7649b4c3e5)

jasmussen commented:
Can work. I wonder if we can tweak the verbiage a bit. Instead of:

> This block is part of the synced pattern "Pattern". To move, delete, or edit other properties, you must edit the pattern. 

perhaps:

> This block belongs to "[pattern-name]". To modify its placement or delete it, edit the pattern."

talldan commented:
> To modify its placement or delete it, edit the pattern."

@jasmussen What do you think about a small tweak for this part? -  "Edit the pattern to modify its placement or delete it."

jameskoster commented:
> Also, I noticed that the "Content" block shows the "Move to" action and can be grabbed but can't actually be moved anywhere. Not sure if this is a bug or not.

Seems like a bug.

> Do we still want the "Create template part" action in the above screenshot?

99% of the time this has very little practical use. I reckon we could try removing it.

> "Edit the pattern to modify its placement or delete it."

Is "modify its placement" a bit verbose? Could be 'This block belongs to "[pattern-name]", edit the pattern to move or delete it'.

kevin940726 committed:
Tweak strings

kevin940726 committed:
Remove convert template menu item if the block is content only

kevin940726 committed:
Fix canMoveBlock with block editing mode

kevin940726 commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1752,14 +1752,16 @@ export function canMoveBlock( state, clientId, rootClientId = null ) {
 	if ( attributes === null ) {
 		return true;
 	}
+	if ( getBlockEditingMode( state, rootClientId ) !== 'default' ) {
```
Move the block editing check earlier so that blocks in the page editor won't be moved (which is expected.)

kevin940726 commented on packages/editor/src/components/block-settings-menu/content-only-settings-menu.native.js:
```diff
@@ -0,0 +1,3 @@
+export default function ContentOnlySettingsMenu() {
```
Don't show in the RN environment. I'm not sure if it's needed though.

talldan replied:
The file this component is used in for the web (packages/editor/src/components/provider/index.js) has a native version of its own that doesn't use the component, so I don't think this is needed.

kevin940726 replied:
Turned out that the native provider is using the universal provider so we still need this.

kevin940726 commented on packages/editor/src/components/block-settings-menu/style.scss:
```diff
@@ -0,0 +1,4 @@
+.editor-content-only-settings-menu__description {
+	padding: $grid-unit;
+	min-width: 235px;
```
An arbitrary width to match the current menu's width with icons.

kevin940726 commented on packages/edit-site/src/components/template-part-converter/index.js:
```diff
@@ -27,12 +27,23 @@ export default function TemplatePartConverter() {
 	);
 
+	// Do not show the convert button if the block is in content-only mode.
+	if ( isContentOnly ) return null;
```
Remove the convert button for blocks in the page editor that's content only.

talldan commented on packages/block-editor/src/components/list-view/block.js:
```diff
@@ -130,8 +130,8 @@ function ListViewBlock( {
-		// Don't show the settings menu if block is disabled or content only.
-		blockEditingMode === 'default';
+		// Don't show the settings menu if block is disabled.
+		blockEditingMode !== 'disabled';
```
I think the whole block wouldn't be shown when disabled, so the check might be unnecessary. Worth looking into it.

talldan commented on packages/editor/src/components/block-settings-menu/content-only-settings-menu.js:
```diff
@@ -0,0 +1,118 @@
+					as="p"
+					className="editor-content-only-settings-menu__description"
+				>
+					{ isPattern
+						? sprintf(
+								// translators: %s: pattern's title.
+								__(
+									'This block belongs to "%s". Edit the pattern to move or delete it.'
+								),
+								entity.title.raw
+						  )
+						: sprintf(
+								// translators: %s: template's title.
+								__(
+									'This block belongs to "%s". Edit the template to move or delete it.'
+								),
+								entity.title.rendered
+						  ) }
```
It is possible to have a `contentOnly` locked group block (or just about any other block) that's not a pattern or a template, and this seems like an unhandled case.

To repro:
1. Add a group block with some inner blocks
2. Go to Code Editor mode and add `"templateLock": "contentOnly"` to the group block's attributes.
3. Return to Visual Editor mode and open the block options menu in List View, observe it mentions being part of a template, even thought that's not the case.

kevin940726 replied:
Is it a case only when the users manually edit the markup in the code editor? I wonder if we should just hide the menu button until we find something to show there? c.c. @jasmussen for any design feedback.

jasmussen replied:
Thanks for the context, and ping. I might need even more context, but let me try and restate my understanding. First, here's a pattern I inserted manually. But I might as well have just typed this out, grouped it, built the pattern myself:
![plain group with stuff](https://github.com/WordPress/gutenberg/assets/1204802/04ce55ee-7158-4ae9-b6bf-81db60dd9897)

If I go into the editor, and add `"templateLock": "contentOnly"` to the group block, it becomes, well, contentOnly locked:

![menu](https://github.com/WordPress/gutenberg/assets/1204802/b765b6f2-8862-4b34-a21d-e9e2f6b29070)

To an extent, this is indeed an unhandled case, because right now there's no way to apply that property to a group without manually editing. IMO that _should_ be possible, because contentOnly locking is a valuable aspect of the editor, something that benefits not just patterns or "sections". As you can see, I'm not seeing that ellipsis menu, though. Did you already hide it, as you suggested, or am I missing a beat?

Hiding it seems fine, at least for now. There's still the larger question: how do I enable contentOnly locking, and how do I turn it off? IMO it could be part of the "Lock" modal, and can be handled separately. CC: @jameskoster

kevin940726 replied:
> As you can see, I'm not seeing that ellipsis menu, though. Did you already hide it, as you suggested, or am I missing a beat?

Yep, I just pushed [8b4ae5e](https://github.com/WordPress/gutenberg/pull/61127/commits/8b4ae5eda2d3c3d2faca8cb276eebb7d227e0856) to hide the menu in this case 😅.

jameskoster replied:
Hmm good discussion.

Hiding may be okay in the short term, but on balance I think I'd lean towards not hiding for a couple of reasons: it gives no explanation about _why_ there's no ellipsis button when you might expect to find one; it's also inconsistent with the other cases.

Couldn't we employ the same pattern we're using in pages / synced patterns and show a contextual message in the menu on click. Something like this, but worded better:

> Only the content of blocks inside "Group" can be edited

There might be a menu item to remove content lock?

kevin940726 replied:
I pushed [4d37b10](https://github.com/WordPress/gutenberg/pull/61127/commits/4d37b1029d2f216980a5d279a905f75fccff209f) to show the suggested description and menu item.

![image](https://github.com/WordPress/gutenberg/assets/7753001/58261ac2-513a-4a01-9265-b6106296ea06)

Upon clicking the "Remove template lock" menu item, the parent (Group in this case) block will remove the "templateLock" attribute. Not sure if it's the desired action.

Not too happy about the wording either 😅. I could use some suggestions. Please feel free to push commits directly too!

jameskoster replied:
Yeah the wording is tricky... I wonder if it would work better the other way around, IE 

> The parent 'Group' block is template-locked, preventing the movement or deletion of child blocks, as well as the addition of any inner blocks.

It's a bit of a mouthful, but maybe that's okay given it's a bit of an edge case at the moment? We probably need a better phrase than "template-locked"... it's not intuitive at all. Maybe...

> The parent 'Group' block is partially locked, preventing the movement or deletion of child blocks, as well as the addition of any inner blocks.
> Unlock 'Group'

kevin940726 replied:
Sure! I pushed [f04744b](https://github.com/WordPress/gutenberg/pull/61127/commits/f04744bebd29baeb1aec5b86f70131b13c3aa18c) for this rewording.

![image](https://github.com/WordPress/gutenberg/assets/7753001/d6719d24-8f81-4095-9609-5ce626f20b16)

jameskoster replied:
I reckon this covers the different cases reasonably well, and certainly improves on the lock icon.

I could go either way on the "Unlock..." menu item. It's a little odd to have a UI to disable a feature, but not one to re-enable it. That could potentially be explored separately.

kevin940726 replied:
> It's a little odd to have a UI to disable a feature, but not one to re-enable it.

Yep, I agree. Though it's also the case for some common actions like "Group" etc. One could always use the Undo button if they really want to.

talldan replied:
We might be able to reuse the existing 'Modify' feature for contentOnly parent blocks.

kevin940726 replied:
In [0147d37](https://github.com/WordPress/gutenberg/pull/61127/commits/0147d37fb8dd8cb55167a33b7f88da8edcdfe978), I updated it to use the "Modify" button as shown in the recordings:


https://github.com/WordPress/gutenberg/assets/7753001/f7379c6a-36e1-425b-b0b2-68748e87648c

talldan commented on packages/editor/src/components/block-settings-menu/content-only-settings-menu.js:
```diff
@@ -0,0 +1,118 @@
+	return (
+		<>
+			<BlockSettingsMenuFirstItem>
+				<Text
+					variant="muted"
+					as="p"
+					className="editor-content-only-settings-menu__description"
+				>
```
This might be something to consider the accessibility of. I don't know how discoverable text would be in a menu like this. It might be fine, or it might be better to also announce the message as an aria-description or something, that's something that might need further feedback.

jameskoster replied:
Good point. This may need to be a more standard popover rather than a `menu`.

jameskoster replied:
@kevin940726 is it feasible to move the text outside of the `role="menu"` container, and give it a different `role`?

kevin940726 replied:
I don't think it's easily achievable 🤔. Do you think it can be done in a follow-up?

talldan reviewed (commented):
This looks great, it's a really nice affordance to explain the various modes to users.

I noticed three things that I think are bugs.

----

One I mentioned in a comment. If you have a contentOnly locked group, the wrong template message shows in a dropdown, which is incorrect. Might need more design feedback on that aspect to know what to show:
![Screenshot 2024-04-30 at 4 40 20 pm](https://github.com/WordPress/gutenberg/assets/677833/f5e20999-8ec1-4bf8-8048-70dea775d776)

----

The second is that when editing a pattern, button blocks with overrides show the 'Add before'/'Add after' buttons, which isn't something that will work:
![Screenshot 2024-04-30 at 4 23 10 pm](https://github.com/WordPress/gutenberg/assets/677833/abaa0caf-92c2-4fe5-85b0-2e08bb534b2d)

----

The third is that when editing a page, blocks like 'Title', 'Featured Image' and 'Content' show the Convert to Template Part option, which I imagine is not desirable and it probably also wouldn't work as those blocks are not replaceable:
![Screenshot 2024-04-30 at 4 50 51 pm](https://github.com/WordPress/gutenberg/assets/677833/f27125a1-7712-4eaa-98b5-73c88d5e4672)

kevin940726 commented:
> The third is that when editing a page, blocks like 'Title', 'Featured Image' and 'Content' show the Convert to Template Part option, which I imagine is not desirable and it probably also wouldn't work as those blocks are not replaceable.

I think I fixed this back in [f824d5a](https://github.com/WordPress/gutenberg/pull/61127/commits/f824d5a726e6fd6469a2016a14cca8b6c005b1a9). Can you still reproduce it?

kevin940726 committed:
Address code review and hide non pattern/template menu

kevin940726 commented:
> If you have a contentOnly locked group, the wrong template message shows in a dropdown, which is incorrect. Might need more design feedback on that aspect to know what to show

> The second is that when editing a pattern, button blocks with overrides show the 'Add before'/'Add after' buttons, which isn't something that will work

These two are fixed in [8b4ae5e](https://github.com/WordPress/gutenberg/pull/61127/commits/8b4ae5eda2d3c3d2faca8cb276eebb7d227e0856).

kevin940726 committed:
Fix lint errors

kevin940726 committed:
Show description for template lock content only

kevin940726 committed:
Fix lint errors

kevin940726 committed:
Reword template lock

kevin940726 committed:
Fix move to still showing in the menu

kevin940726 commented:
The wordings are not final. We can work on that or push a follow-up PR if we haven't decided yet. Both sound good to me!

jameskoster commented:
> **This block belongs to "Pages"**. Edit the template to move or delete it.

This particular example reads a little strange to me. I wonder if this would work better:

> This block is part of the template: "Pages". Edit the template to move or delete it.

Aside from that small tweak I think this is good to try. It feels like a solid improvement on the lock icon to me.

kevin940726 committed:
Change to modify

kevin940726 committed:
Reword the description

kevin940726 commented:
> This block is part of the template: "Pages". Edit the template to move or delete it.

Updated in [a4b4877](https://github.com/WordPress/gutenberg/pull/61127/commits/a4b487732b996b2bf18c2fe45f62d2e6d383905b).

richtabor commented:
How about we simplify a bit, aligning with the Edit/Select mode popover, placing controls before the help text. 

![CleanShot 2024-05-10 at 12 19 44](https://github.com/WordPress/gutenberg/assets/1813435/fabf897d-74b0-4a29-80ff-d6aa895f79a8)

kevin940726 committed:
Reword

kevin940726 commented:
In [a767bd1](https://github.com/WordPress/gutenberg/pull/61127/commits/a767bd1a65a54db24ea72c07d533e405b936b025), I updated the text and removed the references of the block names as suggested in https://github.com/WordPress/gutenberg/pull/61127#issuecomment-2103786568.

**Template Lock: `contentOnly`**
<img width="635" alt="image" src="https://github.com/WordPress/gutenberg/assets/7753001/251ac93a-b8ed-42bb-9cc6-543fb456a0fd">

**Patterns**
<img width="635" alt="image" src="https://github.com/WordPress/gutenberg/assets/7753001/5167193c-0537-4c32-9c94-5091a84277ed">

**Templates**
<img width="635" alt="image" src="https://github.com/WordPress/gutenberg/assets/7753001/842f0aee-fd39-4bba-9dcd-7cb8dd253d95">

talldan commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -391,3 +391,22 @@ export function expandBlock( clientId ) {
 	};
 }
+
+export const modifyContentLockBlock =
```
It'd be good to add a docblock

talldan reviewed (approved):
Couple of small comments, but LGTM. The latest design changes are nice 👍

kevin940726 committed:
Add docblock and delete unused code

kevin940726 committed:
Add back native code

youknowriad commented:
This PR caused a regression in terms of which blocks are locked or not in template-locked mode.

Before:

<img width="351" alt="Screenshot 2024-05-23 at 5 42 20 PM" src="https://github.com/WordPress/gutenberg/assets/272444/80ee0128-52e0-48ec-8f1a-1e7aa6410018">

After:

<img width="349" alt="Screenshot 2024-05-23 at 5 42 49 PM" src="https://github.com/WordPress/gutenberg/assets/272444/5f0a8025-3a63-422c-80d7-6936d7c6d85d">

Mamaduka commented on packages/block-editor/src/components/block-actions/index.js:
```diff
@@ -46,7 +48,9 @@ export default function BlockActions( {
-				canInsertBlock: canInsertDefaultBlock || !! directInsertBlock,
+				canInsertBlock:
+					( canInsertDefaultBlock || !! directInsertBlock ) &&
+					rootBlockEditingMode === 'default',
```
This check turned out to be too restrictive. See #62367.
